File: Hippo/ViewControllers/WebtoonListViewController.m
//
//  WebtoonListViewController.m
//  Hippo
//
//  Created by 전수열 on 13. 9. 1..
//  Copyright (c) 2013년 Joyfl. All rights reserved.
//

#import "WebtoonListViewController.h"
#import "Webtoon.h"
#import "WebtoonDetailViewController.h"
#import "DejalActivityView.h"

@implementation UINavigationBar (Hippo)

- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event
{
	if( !self.clipsToBounds && !self.hidden && self.alpha > 0 )
	{
        for (UIView *subview in self.subviews.reverseObjectEnumerator) {
            CGPoint subPoint = [subview convertPoint:point fromView:self];
            UIView *result = [subview hitTest:subPoint withEvent:event];
            if (result != nil) {
                return result;
                break;
            }
        }
    }
	
    return [super hitTest:point withEvent:event];
}

@end


@implementation WebtoonListViewController


- (void)viewDidLoad
{
	[super viewDidLoad];
	
	self.weekdaySelector = [[WeekdaySelector alloc] init];
	self.weekdaySelector.frame = CGRectMake( 10, 48, UIScreenWidth - 20, self.weekdaySelector.frame.size.height );
	[self.weekdaySelector addTarget:self action:@selector(filterWebtoons) forControlEvents:UIControlEventValueChanged];
	[self.navigationController.navigationBar addSubview:self.weekdaySelector];
	
	self.tableView = [[UITableView alloc] initWithFrame:CGRectMake( 0, 0, UIScreenWidth, UIScreenHeight - 48 )];
	self.tableView.delegate = self;
	self.tableView.dataSource = self;
	self.tableView.contentInset = self.tableView.scrollIndicatorInsets = UIEdgeInsetsMake( 152, 0, 0, 0 );
	[self.view addSubview:self.tableView];
	
	self.searchBar = [[UISearchBar alloc] initWithFrame:CGRectMake( 0, 108, UIScreenWidth, 44 )];
	self.searchBar.placeholder = L( @"SEARCH" );
	self.searchBar.delegate = self;
	[self.view addSubview:self.searchBar];
	
	self.tapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(viewDidTap)];
	
	[[[[self.navigationController.navigationBar.subviews objectAtIndex:0] subviews] objectAtIndex:0] setFrame:CGRectMake( 0, 0, 320, 108 )];
	[[[[self.navigationController.navigationBar.subviews objectAtIndex:0] subviews] objectAtIndex:1] setPosition:CGPointMake( 0, 108 )];
}

- (void)viewWillAppear:(BOOL)animated
{
	if( self.type == HippoWebtoonListViewControllerTypeMyWebtoon ) {
		[self filterWebtoons];
	}
	
	[UIView animateWithDuration:0.25 animations:^{
		self.weekdaySelector.alpha = 1;
		[[[[self.navigationController.navigationBar.subviews objectAtIndex:0] subviews] objectAtIndex:0] setFrame:CGRectMake( 0, 0, 320, 108 )];
		[[[[self.navigationController.navigationBar.subviews objectAtIndex:0] subviews] objectAtIndex:1] setPosition:CGPointMake( 0, 108 )];
	}];
}

- (void)viewDidAppear:(BOOL)animated
{
	self.tableView.contentInset = self.tableView.scrollIndicatorInsets = UIEdgeInsetsMake( 152, 0, 0, 0 );
}

- (void)viewDidDisappear:(BOOL)animated
{
	return;
	[UIView animateWithDuration:0.25 animations:^{
		self.weekdaySelector.alpha = 0;
	} completion:^(BOOL finished) {
		[self.weekdaySelector removeFromSuperview];
	}];
}


#pragma mark -


- (void)filterWebtoons
{
	dispatch_async(dispatch_get_global_queue(0, 0), ^{
		NSString *weekday = HippoWeekdays[self.weekdaySelector.selectedSegmentIndex];
		NSFetchRequest *request = [Webtoon request];
		
		if( self.type == HippoWebtoonListViewControllerTypeMyWebtoon ) {
			request = [request filter:@"subscribed=1"];
		}
		
		if( [weekday isEqualToString:@"finished"] )
		{
			request = [request filter:@"finished=1"];
		}
		else if( ![weekday isEqualToString:@"all"] )
		{
			request = [request filter:@"%@=1 AND finished=0", weekday];
		}
		
		if( self.searchBar.text.length ) {
			request = [request filter:@"title like '*%@*'", self.searchBar.text];
		}
		
		NSString *orderBy = nil;
		if( self.type == HippoWebtoonListViewControllerTypeMyWebtoon ) {
			orderBy = @"new_count desc";
		} else {
			orderBy = @"title";
		}
		
		NSMutableArray *webtoons = [request orderBy:orderBy].all.mutableCopy;
		dispatch_async(dispatch_get_main_queue(), ^{
			self.webtoons = webtoons;
			[self.tableView reloadData];
		});
	});
}


#pragma mark -
#pragma mark UITableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
	return self.webtoons.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
	return 60;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
	static NSString *cellId = @"webtoonCellId";
	WebtoonCell *cell = [tableView dequeueReusableCellWithIdentifier:cellId];
	if( !cell ) {
		cell = [[WebtoonCell alloc] initWithReuseIdentifier:cellId];
		cell.delegate = self;
	}
	
	Webtoon *webtoon = [self.webtoons objectAtIndex:indexPath.row];
	cell.webtoon = webtoon;
	
	return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
	[tableView deselectRowAtIndexPath:indexPath animated:NO];
	
	WebtoonDetailViewController *detailViewController = [[WebtoonDetailViewController alloc] init];
	detailViewController.webtoon = [self.webtoons objectAtIndex:indexPath.row];
	
	self.hidesBottomBarWhenPushed = YES;
	[self.navigationController pushViewController:detailViewController animated:YES];
	self.hidesBottomBarWhenPushed = NO;
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
	if( self.searchBar.isFirstResponder ) {
		[self.searchBar resignFirstResponder];
	}
	
	if( scrollView.contentOffset.y <= -152 )
	{
		self.searchBar.position = CGPointMake( 0, 108 );
	}
	else if( scrollView.contentOffset.y > -152 )
	{
		self.searchBar.position = CGPointMake( 0, -44 - scrollView.contentOffset.y );
	}
}


#pragma mark -
#pragma mark WebtoonCellDelegate

- (void)webtoonCell:(WebtoonCell *)webtoonCell subscribeButtonDidTouchUpInside:(UIButton *)subscribeButton
{
	subscribeButton.enabled = NO;
	
	Webtoon *webtoon = webtoonCell.webtoon;
	if( !webtoon.subscribed.boolValue )
	{
		NSString *api = [NSString stringWithFormat:@"/webtoon/%@/subscribe", webtoon.id];
		[[APILoader sharedLoader] api:api method:@"POST" parameters:nil success:^(id response) {
			subscribeButton.enabled = YES;
			webtoon.subscribed = [NSNumber numberWithBool:YES];
			[JLCoreData saveContext];;
			[webtoonCell layoutContentView];
			
		} failure:^(NSInteger statusCode, NSInteger errorCode, NSString *message) {
			subscribeButton.enabled = YES;
			webtoon.subscribed = [NSNumber numberWithBool:NO];
			[webtoonCell layoutContentView];
		}];
	}
	else
	{
		NSString *api = [NSString stringWithFormat:@"/webtoon/%@/subscribe", webtoon.id];
		[[APILoader sharedLoader] api:api method:@"DELETE" parameters:nil success:^(id response) {
			subscribeButton.enabled = YES;
			webtoon.subscribed = [NSNumber numberWithBool:NO];
			[JLCoreData saveContext];;
			[webtoonCell layoutContentView];
			
		} failure:^(NSInteger statusCode, NSInteger errorCode, NSString *message) {
			subscribeButton.enabled = YES;
			webtoon.subscribed = [NSNumber numberWithBool:YES];
			[webtoonCell layoutContentView];
		}];
	}
}


#pragma mark -
#pragma UISearchBarDelegate

- (BOOL)searchBarShouldBeginEditing:(UISearchBar *)searchBar
{
	[self.tableView addGestureRecognizer:self.tapRecognizer];
	return YES;
}

- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText
{
	[self filterWebtoons];
}

- (void)searchBarSearchButtonClicked:(UISearchBar *)searchBar
{
	[self viewDidTap];
}

- (void)viewDidTap
{
	[self.tableView removeGestureRecognizer:self.tapRecognizer];
	[self.searchBar resignFirstResponder];
}

@end

File: Hippo/ViewControllers/WebtoonListViewController.swift
//
//  WebtoonListViewController.swift
//  Hippo
//
//  Created by 전수열 on 10/3/14.
//  Copyright (c) 2014 Joyfl. All rights reserved.
//

import Snappy
import UIKit

enum WebtoonListType {
    case Mine, All
}

class WebtoonListViewController: UIViewController, UISearchBarDelegate, UITableViewDataSource, UITableViewDelegate {

    private struct CellID {
        static let Webtoon = "Webtoon"
    }

    private var _listType = WebtoonListType.All
    var listType: WebtoonListType {
        get {
            return _listType
        }
        set {
            _listType = newValue
            self.title = newValue == WebtoonListType.All ? __("Search") : __("My Webtoons")
        }
    }

    let weekdaySelector = UISegmentedControl()
    let searchBar = UISearchBar()
    let tableView = UITableView()
    let tapRecognizer = UITapGestureRecognizer()

    var webtoons: RLMArray?

    override func viewDidLoad() {
        super.viewDidLoad()

        self.tableView.registerClass(WebtoonCell.self, forCellReuseIdentifier: CellID.Webtoon)
        self.tableView.dataSource = self
        self.tableView.delegate = self
        self.tableView.contentInset = UIEdgeInsetsMake(88, 0, 0, 0)
        self.tableView.scrollIndicatorInsets = self.tableView.contentInset
        self.view.addSubview(self.tableView)
        self.tableView.snp_makeConstraints { make in
            make.size.equalTo(self.view)
            return
        }

        self.searchBar.delegate = self
        self.searchBar.placeholder = __("Search")
        self.view.addSubview(self.searchBar)

        let toolbar = UIToolbar()
        self.view.addSubview(toolbar)
        toolbar.snp_makeConstraints { make in
            make.top.equalTo(64)
            make.width.equalTo(self.view)
        }

        toolbar.addSubview(self.weekdaySelector)
        let options = ["All", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun", "Con"]
        for i in 0...(options.count - 1) {
            self.weekdaySelector.insertSegmentWithTitle(__(options[i]), atIndex: i, animated: false)
        }
        self.weekdaySelector.addTarget(self, action: "filterWebtoons", forControlEvents: UIControlEvents.ValueChanged)
        self.weekdaySelector.snp_makeConstraints { make in
            make.centerX.equalTo(toolbar)
            make.centerY.equalTo(toolbar).with.offset(-1)
            make.width.equalTo(toolbar).with.offset(-20)
            return
        }

        self.searchBar.snp_makeConstraints { make in
            make.top.equalTo(toolbar.snp_bottom)
            make.width.equalTo(self.view)
            make.height.equalTo(44)
        }

        self.tapRecognizer.addTarget(self, action: "viewDidTap")
    }

    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)

        if self.weekdaySelector.selectedSegmentIndex == UISegmentedControlNoSegment {
            let calendar = NSCalendar.currentCalendar()
            let components = calendar.components(NSCalendarUnit.WeekdayCalendarUnit, fromDate: NSDate())
            let weekday = components.weekday
            self.weekdaySelector.selectedSegmentIndex = (weekday >= 2 ? weekday - 1 : 7)
        }

        filterWebtoons()
    }

    // MARK: - Filter

    func filterWebtoons() {
        let options = ["all", "mon", "tue", "wed", "thu", "fri", "sat", "sun", "concluded"]
        let option = options[self.weekdaySelector.selectedSegmentIndex]

        var predicate = ""
        if self.listType == WebtoonListType.Mine {
            predicate += "subscribing=1"
        }
        if self.weekdaySelector.selectedSegmentIndex > 0 {
            if predicate != "" {
                predicate += " AND "
            }
            predicate += "\(option)=1"
        }

        if !self.searchBar.text.isEmpty {
            if predicate != "" {
                predicate += " AND "
            }
            predicate += "title CONTAINS[c] '\(self.searchBar.text)'"
        }

        if predicate == "" {
            self.webtoons = Webtoon.allObjects()
        } else {
            self.webtoons = Webtoon.objectsWithPredicate(NSPredicate(format: predicate))
        }
        self.tableView.reloadData()
    }

    // MARK: - UITableViewDeleagate

    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if self.webtoons == nil {
            return 0
        }
        return Int(self.webtoons!.count)
    }

    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return WebtoonCell.height
    }

    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier(CellID.Webtoon) as? WebtoonCell
        cell?.webtoon = self.webtoons?.objectAtIndex(UInt(indexPath.row)) as? Webtoon
        return cell!
    }

    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated: false)

        let detailViewController = WebtoonDetailViewController()
        detailViewController.webtoon = self.webtoons!.objectAtIndex(UInt(indexPath.row)) as? Webtoon
        self.navigationController?.pushViewController(detailViewController, animated: true)
    }

    func scrollViewDidScroll(scrollView: UIScrollView) {
        if self.searchBar.isFirstResponder() {
            self.searchBar.resignFirstResponder()
        }

        if scrollView.contentOffset.y <= -152 {
            self.searchBar.y = 108
        } else if scrollView.contentOffset.y > -152 {
            self.searchBar.y = -44 - scrollView.contentOffset.y
        }
    }

    // MARK: - UISearchBarDelegate

    func searchBarShouldBeginEditing(searchBar: UISearchBar) -> Bool {
        self.tableView.addGestureRecognizer(self.tapRecognizer)
        return true
    }

    func searchBar(searchBar: UISearchBar, textDidChange searchText: String) {
        self.filterWebtoons()
    }

    func searchBarSearchButtonClicked(searchBar: UISearchBar) {
        self.viewDidTap()
    }

    // MARK: - UITapGestureRecognizer

    func viewDidTap() {
        self.tableView.removeGestureRecognizer(self.tapRecognizer)
        self.searchBar.resignFirstResponder()
    }
}
